specifications/KeyValueStore/ClientCentric.tla

```diff
@@ -0,0 +1,185 @@
+\* Note: This execution does therefor not contain the "final state" of the execution, since it is not a parent state of a transaction.
+Execution == Seq(ExecutionElem)
+
+\* Seq
+executionStates(execution) == [ i \in 1..Len(execution) |-> execution[i].parentState ]
+\* Set
+executionTransactions(execution) == { ep.transaction : ep \in SeqToSet(execution) }
+
+\* "The parent state is the last state in the execution
+\* Definition 1: s -T-> s' ≡ [(k,v) ∈ s ∧ (k,v)􏰍 \notin s'] => k ∈ W_T /\ w(k,v) ∈ Σ_T => (k,v) ∈ s.
+\* We refer to s as the parent state of T (denoted as sp,T ); to the
+\* transaction that generated s as Ts ; and to the set of keys in which
+\* s and s′ differ as ∆(s,s′)"
+parentState(execution, transaction) == 
+  LET ind == CHOOSE i \in 1..Len(execution): execution[i].transaction = transaction
+  IN execution[ind].parentState
+
+\* w(k,v) -to-> r(k,v)
+\* check reads and writes, implicit because of "write" check in ReadStates
+earlierInTransaction(transaction, op1, op2) == Index(transaction, op1) < Index(transaction, op2)
+
+\* state1 -*-> state2
+beforeOrEqualInExecution(execution, state1, state2) == 
+  LET states == executionStates(execution)
+  IN  Index(states, state1) <= Index(states, state2)
+
+\* Read states: from which states can the operation read it's value
+ReadStates(execution, operation, transaction) == 
+  LET Se == SeqToSet(executionStates(execution))
+      sp == parentState(execution, transaction)
+  IN { s \in Se:
+        /\ beforeOrEqualInExecution(execution, s, sp) \* s -*-> s_p: restrict the valid read states for the operations in T to be no later than sp
+        /\ \/ s[operation.key] = operation.value \* (k,v) \in s
+           \/ \E write \in SeqToSet(transaction): 
+              /\ write.op = "write" /\ write.key = operation.key /\ write.value = operation.value
+              /\ earlierInTransaction(transaction, write, operation) \* w(k,v)-to->r(k,v)
+\* "By convention, write operations have read states too: for a write operation in T , they include all states in Se up to and including T ’s parent state."
+           \/ operation.op = "write"
+     }
+        
+Preread(execution, transaction) ==
+  \A operation \in SeqToSet(transaction): ReadStates(execution, operation, transaction) /= {}
+
+PrereadAll(execution, transactions) == 
+  \A transaction \in transactions: Preread(execution, transaction)
+
+\* A state `s` is complete for `T` in `e` if every operation in `T` can read from `s`
+Complete(execution, transaction, state) == 
+  LET setOfAllReadStatesOfOperation(transactionOperations) ==
+        { ReadStates(execution, operation, transaction) : operation \in SeqToSet(transactionOperations) }
+     \* also include all states for when the transaction contains no operations
+      readStatesForEmptyTransaction == { s \in SeqToSet(executionStates(execution)) : beforeOrEqualInExecution(execution, s, parentState(execution, transaction)) }
```

Comment:
I don't see how this definition (or its use) is restricted to the case "when the transaction contains no operation"? Perhaps rephrase the comment?

Reply:
ClientCentric is not my module, written by Tim, so I am reluctant to make any change there.

Reply:
/cc @TimSoethout

Reply:
If I remember correctly, this is where I deviated from the original definitions in the paper by Crooks (https://www.cs.cornell.edu/lorenzo/papers/Crooks17Seeing.pdf). The comment is mainly there to upgrade the definition to also allow transactions with an empty operation set, while still satisfying the isolation level. 
I personally used when incrementally model checking isolation levels of the Two-Phase Locking and Local-Coordination Avoidance protocols, which contained empty transactions when initializing.
The comment could for example be rephrased to: 
"`readStatesForEmptyTransaction` contains all previous states, in order to make sure that empty transactions do not incorrectly invalidate the checked isolation level."

Reply:
Has this conversation been resolved?  If yes, it looks like this PR is ready to be merged.

Reply:
I certainly have no objection to merging the PR, although I don't think that the wording of the comment has been changed (Tim's suggestion clarifies the intent of the definition IMHO).

Reply:
@muratdem can you do the finishing touches (wording of this comment + removal `[]` copy/paste artifact), and the we merge it?

Reply:
yes, I will also pull some global variables to be process local.